File: src/ros2_bridge/mappers/foxglove_msgs/odometry.rs
//! Typed mapper for `foxglove_msgs/msg/Odometry`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn odometry_to_bus(msg: ros_env::foxglove_msgs::msg::Odometry) -> crate::foxglove_msgs::msg::v1::Odometry {
    crate::foxglove_msgs::msg::v1::Odometry {
        timestamp: Some(crate::ros2_bridge::mappers::convert::time_to_timestamp(msg.timestamp)),
        frame_id: crate::ros2_bridge::mappers::convert::from_ros_string(msg.frame_id),
        body_frame_id: crate::ros2_bridge::mappers::convert::from_ros_string(msg.body_frame_id),
        pose: Some(crate::ros2_bridge::mappers::foxglove_msgs::pose::pose_to_bus(msg.pose)),
        linear_velocity: msg.linear_velocity.map(crate::ros2_bridge::mappers::foxglove_msgs::vector3::vector3_to_bus),
        angular_velocity: msg.angular_velocity.map(crate::ros2_bridge::mappers::foxglove_msgs::vector3::vector3_to_bus),
        pose_covariance: crate::ros2_bridge::mappers::convert::f64_seq(msg.pose_covariance),
        velocity_covariance: crate::ros2_bridge::mappers::convert::f64_seq(msg.velocity_covariance),
        metadata: msg.metadata.into_iter().map(crate::ros2_bridge::mappers::foxglove_msgs::key_value_pair::key_value_pair_to_bus).collect(),
    }
}

pub(crate) fn odometry_to_ros(bus: crate::foxglove_msgs::msg::v1::Odometry) -> ros_env::foxglove_msgs::msg::Odometry {
    ros_env::foxglove_msgs::msg::Odometry {
        timestamp: crate::ros2_bridge::mappers::convert::timestamp_to_time(bus.timestamp.unwrap_or_default()),
        frame_id: crate::ros2_bridge::mappers::convert::to_ros_string(bus.frame_id),
        body_frame_id: crate::ros2_bridge::mappers::convert::to_ros_string(bus.body_frame_id),
        pose: crate::ros2_bridge::mappers::foxglove_msgs::pose::pose_to_ros(bus.pose.unwrap_or_default()),
        linear_velocity: bus.linear_velocity.map(crate::ros2_bridge::mappers::foxglove_msgs::vector3::vector3_to_ros),
        angular_velocity: bus.angular_velocity.map(crate::ros2_bridge::mappers::foxglove_msgs::vector3::vector3_to_ros),
        pose_covariance: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.pose_covariance),
        velocity_covariance: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.velocity_covariance),
        metadata: bus.metadata.into_iter().map(crate::ros2_bridge::mappers::foxglove_msgs::key_value_pair::key_value_pair_to_ros).collect(),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct FoxgloveMsgsOdometryMapper;

impl TypedTopicMapper for FoxgloveMsgsOdometryMapper {
    type Ros = ros_env::foxglove_msgs::msg::Odometry;
    type Bus = crate::foxglove_msgs::msg::v1::Odometry;

    fn type_name(&self) -> &'static str {
        "foxglove_msgs/msg/Odometry"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(odometry_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(odometry_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/foxglove_msgs/odometry.py
"""Generated mapper for `foxglove_msgs/msg/Odometry`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.foxglove_msgs.pose import pose_to_bus, pose_to_ros
from robot_bus.ros2_bridge.mappers.foxglove_msgs.vector3 import vector3_to_bus, vector3_to_ros
from robot_bus.ros2_bridge.mappers.foxglove_msgs.key_value_pair import key_value_pair_to_bus, key_value_pair_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.foxglove_msgs.msg.v1 import Odometry as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def odometry_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.timestamp = _convert.time_to_timestamp(msg.timestamp)
    bus.frame_id = str(msg.frame_id)
    bus.body_frame_id = str(msg.body_frame_id)
    bus.pose.CopyFrom(pose_to_bus(msg.pose))
    bus.linear_velocity.CopyFrom(vector3_to_bus(msg.linear_velocity))
    bus.angular_velocity.CopyFrom(vector3_to_bus(msg.angular_velocity))
    bus.pose_covariance.extend(list(msg.pose_covariance))
    bus.velocity_covariance.extend(list(msg.velocity_covariance))
    bus.metadata.extend([key_value_pair_to_bus(x) for x in msg.metadata])
    return bus


def odometry_to_ros(bus):
    from foxglove_msgs.msg import Odometry as RosMsg

    out = RosMsg()
    out.timestamp = _convert.timestamp_to_time(bus.timestamp)
    out.frame_id = str(bus.frame_id)
    out.body_frame_id = str(bus.body_frame_id)
    out.pose = pose_to_ros(bus.pose)
    out.linear_velocity = vector3_to_ros(bus.linear_velocity)
    out.angular_velocity = vector3_to_ros(bus.angular_velocity)
    out.pose_covariance = list(bus.pose_covariance)
    out.velocity_covariance = list(bus.velocity_covariance)
    out.metadata = [key_value_pair_to_ros(x) for x in bus.metadata]
    return out


class FoxgloveMsgsOdometryMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from foxglove_msgs.msg import Odometry as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return odometry_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return odometry_to_ros(bus)
